backend/headless/output.c

```diff
@@ -120,6 +120,23 @@ static void output_rollback_render(struct wlr_output *wlr_output) {
 	output->back_buffer = NULL;
 }
 
+static bool output_export_dmabuf(struct wlr_output *wlr_output,
+		struct wlr_dmabuf_attributes *attribs) {
+	struct wlr_headless_output *output =
+		headless_output_from_output(wlr_output);
+
+	if (!output->front_buffer) {
+		return false;
+	}
+
+	struct wlr_dmabuf_attributes tmp;
+	if (!wlr_buffer_get_dmabuf(output->front_buffer, &tmp)) {
```

Comment:
Should this not be `back_buffer`? This works well with wayvnc if the back-buffer is used.

Reply:
The DRM backend [exports the front-buffer too](https://github.com/swaywm/wlroots/blob/297354f8477285231826f6441b943188ec7d7399/backend/drm/drm.c#L642), so I just went with that. But the [wlr_output docs](https://github.com/swaywm/wlroots/blob/297354f8477285231826f6441b943188ec7d7399/include/wlr/types/wlr_output.h#L389) seem to not agree.

Reply:
wlr-screencopy expects the back-buffer and wlr-export-dmabuf seems to also expect it. What should be the source of truth here?

Reply:
That's a good question.

I think the original behaviour was to hand over the back buffer. Then this commit changed it to use the front-buffer: https://github.com/swaywm/wlroots/commit/8da9d9679e4cad640b0f0a518ca1529c8cdecbe0#diff-8b05a774317ee8e87d51422170f82d4bL663

However, I think moving forward it'd make more sense to hand over the front buffer. Acquiring the back-buffer has a number of limitations: there's no guarantee we even have a back buffer, except right before a commit. I'd like to move away from implicit state like this.

This would involve updating the docs + screencopy/export-dmabuf to export on `commit` instead of `precommit`.

Reply:
Can you make `glReadPixels` read the front buffer without implementing this PR for all backends?

Reply:
Unfortunately, it's not possible for GLES2 to select the front buffer. `glReadBuffer(GL_FRONT)` is only available on desktop GL.

Reply:
There are two options, right?
 1. Have drm read from the back-buffer and keep this PR as is, and slowly work towards the goal of implementing this for all the backends, one backend at a time. Then, eventually, switch to reading from and exporting the front buffer.
 2. Implement this PR for all backends and change screencopy and export-dmabuf to use the front-buffer in one big PR.

Reply:
Submitted https://github.com/swaywm/wlroots/pull/2448 to do (1).

Reply:
>Submitted #2448 to do (1).

This PR is toast: we don't have access to the back-buffer prior to the output commit: https://github.com/swaywm/wlroots/blob/79c7591dc196b14dfa857caf2c183276f937ee02/backend/drm/renderer.c#L282

>2. Implement this PR for all backends and change screencopy and export-dmabuf to use the front-buffer in one big PR.

Actually, we don't need to read from the front-buffer when using the `glReadPixels` code-path. We can continue reading from the back-buffer for shm, *and* read from the front-buffer for DMA-BUF. This will probably create friction for screencopy which supports both, but I'll see if it's manageable.

Reply:
Whoops. :)

I suspect that the damage tracking is going to be the hardest part.

Reply:
>We can continue reading from the back-buffer for shm, and read from the front-buffer for DMA-BUF.

I've done that here: https://github.com/swaywm/wlroots/pull/2463